sklearn/utils/testing.py

```diff
@@ -393,6 +403,177 @@ def assert_raise_message(exception, message, function, *args, **kwargs):
+                attr.endswith('estimators_') or
+                type(value1) not in comparables or
+                attr in skip_attributes):
+            continue
+
+        # If the attr is not present in the 2nd estimator
+        if attr not in est2_dict:
+            return False, 'attr_not_in_est2', attr
+
+        value2 = est2_dict[attr]
+        value1_type = type(value1)
+        value2_type = type(value2)
+
+        # If the attr types differ between the two estimators
+        if value1_type is not value2_type:
+            return (False, 'attr_not_same_type',
+                    (value1, value1_type), (value2, value2_type))
+
+        # TODO Support sparse matrices?
+        # If the attribute is of list / ndarray type
+        if value1_type in (np.ndarray, list):
+            # Check if the content(s) of the list are comparable and
+            # that the list is not empty
+            try:
+                if (type(value1[0]) not in comparables
+                        or type(value2[0]) not in comparables):
+                    continue
+            except IndexError:
+                # If the list / ndarray is empty
+                pass
+            if ((np.shape(value1) != np.shape(value2)) or
+                    not np.allclose(value1, value2)):
+                return False, 'attr_values_differ', attr, value1, value2
+            if not np.allclose(value1, value2):
+                return False, 'attr_values_differ', attr, value1, value2
+
+        # If the attribute is of numerical type
+        if value1_type in (float, np.float64, int, 'str'):
+            if not value1 == value2:
+                return False, 'attr_values_differ', attr, value1, value2
+    return (True,)
+
+
+def assert_fitted_attributes_equal(estimator1, estimator2):
+    """Helper function to check if fitted model attributes are equal"""
+    debug_msg = ""
+    result = _attributes_equal(estimator1, estimator2)
+    if not result[0]:
+        if result[1] == 'attr_not_in_est2':
+            debug_msg = ('Attribute, %s, was not found in estimator 2'
+                         % result[2])
+        elif result[1] == 'attr_not_same_type':
+            debug_msg = ('Attribute, %s, is %r with type %r in estimator1'
+                         ' and is %r with type %r in estimator2'
+                         % ((result[2],) + result[3] + result[4]))
+        elif result[1] == 'attr_values_differ':
+            debug_msg = ('Attribute, %s, differs between the two estimators.'
+                         ' The differing values are %r and %r'
+                         % result[2:])
+    assert_true(result[0],
+                "The attributes differ between models.\n" + debug_msg)
+
+
+def assert_fitted_attributes_not_equal(estimator1, estimator2):
+    """Helper function to check if fitted model attributes are not equal"""
+    assert_false(_attributes_equal(estimator1, estimator2)[0],
+                 "The attributes do not differ between models.")
+
+
```

Comment:
@amueller @ogrisel does this look better... or do you feel I could simplify this even more?